src/enzyme_ad/jax/Passes/LowerJIT.cpp

```diff
@@ -989,12 +989,15 @@ struct LowerJITPass
           NamedAttribute(rewriter.getStringAttr("attr"), backendstr));
       auto dattr = DictionaryAttr::get(op.getContext(), names);
 
+      BoolAttr hasSideEffectAttr =
+          rewriter.getBoolAttr(!isMemoryEffectFree(op));
```

Comment:
so this is too conservative here. In particular as it currently is nearly everything will be marked as side-effecting (even if the code doesn't need to). Specifically here we only need to mark things as side-effective if it could write to data which isn't a function argument (e.g. if we were to write to a function argument, that would be represented as an updated result perfectly fine).

Reply:
I'd say let's start conservative and then relax progressively. If we go into the per-value effect territory (the representation supports both and arguably there are "global" effects without value that we may not want to ignore, e.g. synchronizations), we may need to account for aliasing, at which point we should just borrow the enzyme-mlir analyses.

Reply:
yeah that's fine, but it does mean we need an intermediate attribute to force it to be side effect free (current lowerings ought actually be side effect free but are not deduced as such).